### PR TITLE
Introduce exception handling support for lifecycle callback methods

### PR DESCRIPTION
**Pull request by phoenix384**
## Overview
PR for Issue #1454 
Make it possible to also handle exceptions thrown by BeforeEach/AfterEach methods via the TestExecutionExceptionHandler.
The new handler methods are default methods to make them optional and to not break the FunctionalInterface.
This PR is a proposal to solve the issue. It's not complete. The Javadoc is not adjusted and the exception handling for the BeforeAll/AfterAll methods is missing.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [ ] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [ ] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by phoenix384** [comment]
Added exception handling for BeforeAll/AfterAll methods

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassTestDescriptor.java [hunk 1]
```diff
@@ -358,9 +362,16 @@ private void invokeBeforeAllMethods(JupiterEngineExecutionContext context) {
 		ThrowableCollector throwableCollector = context.getThrowableCollector();
 		Object testInstance = extensionContext.getTestInstance().orElse(null);
 
-		for (Method method : this.beforeAllMethods) {
-			throwableCollector.execute(
-				() -> executableInvoker.invoke(method, testInstance, extensionContext, registry));
+		for (Method method: this.beforeAllMethods) {
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassTestDescriptor.java]: Please reintroduce the space between `method` and `:`.

Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/ClassTestDescriptor.java]: Please reintroduce the space between `method` and `:`.

**Comment by sbrannen** [comment]
The build is currently broken due to formatting issues.

> Run `gradlew spotlessApply` to fix these violations.

**Comment by phoenix384** [comment]
But where to put the tests (TestExecutionExceptionHandlerTests or BeforeAndAfterEachTests or a new class) without blowing up an existing class or repeating too much code in a new class?

**Comment by sbrannen** [comment]
I'd probably just introduce a set of test classes parallel to the existing `TestExecutionExceptionHandlerTests`.

As for adhering to the DRY principle, just do your best. We can always refactor later. The main objective for starters is just having tests in place.

**Comment by sbrannen** [comment]
and yeah.... please don't "blow up" anything. 😉

**Comment by phoenix384** [comment]
Tests added

**Comment by phoenix384** [comment]
Finally removed the formatting issues - these checks are _really_ pedantic...
Don't know why there is still a failure.

**Comment by jbduncan** [comment]
@phoenix384 Were you not able to simply run `gradlew spotlessApply` to fix all the formatting problems? :thinking:

**Comment by sbrannen** [comment]
@phoenix384, you really shouldn't try to tackle formatting issues one-by-one.

Instead, as @jbduncan mentioned, you should simply execute `gradlew spotlessApply` once before committing your changes.

Alternatively, you can set up the formatting in your IDE.

This is all documented in the [CONTRIBUTING](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#formatting) documentation for JUnit 5.

**Comment by sbrannen** [comment]
The last failure appears to have something to do with Jacoco:

https://travis-ci.org/junit-team/junit5/jobs/413706488#L1467-L1486

Maybe that's just an intermittent issue on the CI server.

Does `gradle build` work consistently on your machine?

**Comment by phoenix384** [comment]
5.4 is released and this is still not merged. Is there a reason?

**Comment by sbrannen** [comment]
Hi @phoenix384,

> 5.4 is released and this is still not merged. Is there a reason?

Yes, requested changes have not yet been made (e.g., https://github.com/junit-team/junit5/pull/1509#discussion_r217681089), and the following tasks under "Definition of Done" have not been addressed.

- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [ ] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

Please rebase on `master` and address those 3 tasks, and we'll make sure to review it again and get it into JUnit 5.5.

Thanks!

**Comment by marcphilipp** [comment]
Closing in favor of #1868.

**Pull request closed without merging**